File: mobile/engine/src/main/kotlin/com/branchconsole/engine/transforms/Transforms.kt
package com.branchconsole.engine.transforms

import kotlin.math.abs

/**
 * `engine_ref/transforms.py` 1:1 이식 (docs/plans/M1_PLAN_A.md §2.4.2 체크리스트,
 * M1_PLAN_D.md §2.4 전치 정리). 롤링 집계가 필요한 나머지 함수는 [RollingTransforms]에 있다
 * (같은 규율을 공유하는 한 몸이며, 파일 분리는 순수히 detekt `TooManyFunctions` 완화 목적).
 *
 * 규율:
 *  - K-07: 전부 Double(float64) 고정, 반올림 없음(표시 계층 몫).
 *  - rolling 계열은 전부 `min_periods = window`다 — 창 안에 **정확히 window개의 non-NaN 값이
 *    연속으로** 있어야 값을 낸다(하나라도 NaN이면 그 행은 NaN). 앞쪽 `window-1`행은 항상 NaN
 *    (pandas `Series.rolling(window, min_periods=window)` 규약, 파리티 지뢰 2).
 *  - `Double.NaN`이 결측 표식이다. IEEE-754 비교 규약(NaN과의 모든 비교는 false)이 pandas/
 *    numpy NaN 산술과 동일하게 성립한다 — 별도 null 래핑을 쓰지 않는다. 이 성질을 지키려면
 *    호출부는 이 파일의 `Double`을 절대 박싱된 `Double?`로 바꾸면 안 된다(`Double?`의 `==`는
 *    NaN==NaN을 true로 취급해 이 규약을 깬다).
 *  - 표준편차는 ddof=1(표본) — pandas `rolling().std()` 기본값(파리티 지뢰 1).
 *  - 순수 함수, 부작용 없음. window·lookback 등 파라미터는 전부 호출부
 *    ([com.branchconsole.engine.config.TransformParser])에서 주입받는다 — 숫자 리터럴 금지
 *    (CLAUDE.md §1).
 */
object Transforms {
    private const val PERCENT = 100.0
    private const val PERIODS_1D = 1
    private const val PERIODS_5D = 5

    /** rolling z-score. absolute=true면 절대값 변형(dxy_z 등 direction=higher_is_risk용). */
    fun zscore(
        x: DoubleArray,
        window: Int,
        absolute: Boolean = false,
    ): DoubleArray {
        val mean = RollingWindow.mean(x, window)
        val std = RollingWindow.std(x, window)
        return DoubleArray(x.size) { i ->
            val z = (x[i] - mean[i]) / std[i]
            if (absolute) abs(z) else z
        }
    }

    /** 두 배열이 이미 같은(as_of) 인덱스로 정렬돼 있다고 가정한다(포지션 기반) — 서로 다른
     * 인덱스의 정렬은 [SeriesAlign.unionAlign]의 몫(pandas 자동 인덱스 정렬 대응). */
    fun ratio(
        a: DoubleArray,
        b: DoubleArray,
    ): DoubleArray {
        require(a.size == b.size) { "ratio: size mismatch (${a.size} vs ${b.size}) — align first" }
        return DoubleArray(a.size) { i -> a[i] / b[i] }
    }

    /** %-단위 시계열의 lookback일 변화를 bp로(1%p = 100bp). */
    fun deltaBp(
        x: DoubleArray,
        lookback: Int,
    ): DoubleArray = DoubleArray(x.size) { i -> if (i < lookback) Double.NaN else (x[i] - x[i - lookback]) * PERCENT }

    fun pctChange(
        x: DoubleArray,
        periods: Int,
    ): DoubleArray =
        DoubleArray(x.size) { i ->
            if (i < periods) Double.NaN else (x[i] - x[i - periods]) / x[i - periods] * PERCENT
        }

    fun pctChange1d(x: DoubleArray): DoubleArray = pctChange(x, PERIODS_1D)

    fun pctChange5d(x: DoubleArray): DoubleArray = pctChange(x, PERIODS_5D)

    fun absValue(x: DoubleArray): DoubleArray = DoubleArray(x.size) { i -> abs(x[i]) }

    /** 롤링 고점 대비 낙폭 %(양수 = 하락, 0 = 신고가). */
    fun drawdownFromHigh(
        x: DoubleArray,
        window: Int,
    ): DoubleArray {
        val rollingHigh = RollingWindow.max(x, window)
        return DoubleArray(x.size) { i -> (rollingHigh[i] - x[i]) / rollingHigh[i] * PERCENT }
    }

    /** 하락(음의 변화)이 위험(+)이 되도록 부호 반전한 z-score. */
    fun negZscore(
        x: DoubleArray,
        window: Int,
    ): DoubleArray {
        val z = zscore(x, window)
        return DoubleArray(z.size) { i -> -z[i] }
    }
}

File: engine_ref/transforms.py
"""engine_ref.transforms — pandas Series 기반 순수 변환 함수 (D-01 활성 지표 15종이 요구하는 전부).

규율(CLAUDE.md §1): 이 모듈은 숫자를 모른다. window·lookback 등 파라미터는 전부 호출부
(engine_ref.registry가 configs/indicators.yaml의 transform 문자열을 파싱한 값)에서 주입받는다.
전부 float64 고정(K-07 — 반올림은 표시 계층 몫), 부작용 없는 순수 함수.
"""

from __future__ import annotations

import operator

import numpy as np
import pandas as pd

_OPS = {
    "<": operator.lt,
    "<=": operator.le,
    ">": operator.gt,
    ">=": operator.ge,
    "==": operator.eq,
}


def zscore(x: pd.Series, window: int, *, absolute: bool = False) -> pd.Series:
    """rolling z-score. absolute=True면 절대값 변형(dxy_z 등 direction=higher_is_risk용)."""
    x = x.astype("float64")
    mean = x.rolling(window, min_periods=window).mean()
    std = x.rolling(window, min_periods=window).std()
    z = (x - mean) / std
    return z.abs() if absolute else z


def ratio(a: pd.Series, b: pd.Series) -> pd.Series:
    return a.astype("float64") / b.astype("float64")


def delta_bp(x: pd.Series, lookback: int) -> pd.Series:
    """% 단위 시계열의 lookback일 변화를 bp로 (1%p = 100bp)."""
    x = x.astype("float64")
    return (x - x.shift(lookback)) * 100.0


def pct_change_1d(x: pd.Series) -> pd.Series:
    return x.astype("float64").pct_change(1, fill_method=None) * 100.0


def pct_change_5d(x: pd.Series) -> pd.Series:
    return x.astype("float64").pct_change(5, fill_method=None) * 100.0


def abs_(x: pd.Series) -> pd.Series:
    return x.astype("float64").abs()


def drawdown_from_high(x: pd.Series, window: int) -> pd.Series:
    """롤링 고점 대비 낙폭 %(양수 = 하락, 0 = 신고가)."""
    x = x.astype("float64")
    rolling_high = x.rolling(window, min_periods=window).max()
    return (rolling_high - x) / rolling_high * 100.0


def neg_zscore(x: pd.Series, window: int) -> pd.Series:
    """하락(음의 변화)이 위험(+)이 되도록 부호 반전한 z-score."""
    return -zscore(x, window)


def rolling_corr(a: pd.Series, b: pd.Series, window: int) -> pd.Series:
    return (
        a.astype("float64")
        .rolling(window, min_periods=window)
        .corr(b.astype("float64"))
    )


def rolling_mean_corr(corr: pd.Series, window: int) -> pd.Series:
    """장기 평균 상관(예: mean_corr120) — corr 시계열 자체의 롤링 평균."""
    return corr.astype("float64").rolling(window, min_periods=window).mean()


def rolling_sum(x: pd.Series, window: int) -> pd.Series:
    return x.astype("float64").rolling(window, min_periods=window).sum()


def gate_mask(series: pd.Series, op: str, threshold: float) -> pd.Series:
    """게이트 조건 불리언 마스크. op는 engine_ref.registry.parse_gate가 파싱한 비교 연산자."""
    return _OPS[op](series.astype("float64"), threshold)


def gated(z: pd.Series, mask: pd.Series) -> pd.Series:
    """mask가 False인 지점을 0으로 마스킹(결측 아님 — severity 0으로 이어짐). z·mask는 같은 인덱스 전제."""
    return z.astype("float64").where(mask, 0.0)


def realized_vol_kospi_20d(daily_return_pct: pd.Series, window: int) -> pd.Series:
    """K-02 VKOSPI 폴백: KOSPI 일수익률(%) N일 실현변동성 연율화(%).

    window은 하드코딩하지 않는다 — 호출부가 폴백 식별자 문자열
    (source.fallback: "realized_vol_kospi_20d")에서 engine_ref.registry.parse_fallback_window로
    파싱해 주입한다. 이 출력에 이후 zscore(window=252)를 적용하는 것은 registry/scoring 계층의 몫이다.
    """
    r = daily_return_pct.astype("float64") / 100.0
    return r.rolling(window, min_periods=window).std() * np.sqrt(252) * 100.0
